packages/salesforcedx-vscode-core/src/commands/forceDataSoqlQuery.ts

```diff
@@ -17,59 +17,98 @@ import {
 import * as vscode from 'vscode';
 import { nls } from '../messages';
 import {
+  CompositeParametersGatherer,
   SfdxCommandlet,
   SfdxCommandletExecutor,
   SfdxWorkspaceChecker
 } from './commands';
 
 class ForceDataSoqlQueryExecutor extends SfdxCommandletExecutor<{}> {
-  public build(data: QueryInput): Command {
-    return new SfdxCommandBuilder()
+  public build(data: QueryAndApi): Command {
+    let command = new SfdxCommandBuilder()
       .withDescription(nls.localize('force_data_soql_query_input_text'))
       .withArg('force:data:soql:query')
-      .withFlag('--query', `${data.input}`)
-      .build();
+      .withFlag('--query', `${data.query}`);
+    if (data.api === ApiType.Tooling) {
+      command = command.withArg('--usetoolingapi');
+    }
+    return command.build();
   }
 }
 
-export type QueryInput = {
-  input: string;
-};
-
-export class GetQueryInput implements ParametersGatherer<{ input: string }> {
+export class GetQueryInput implements ParametersGatherer<QueryInput> {
   public async gather(): Promise<
-    CancelResponse | ContinueResponse<{ input: string }>
+    CancelResponse | ContinueResponse<QueryInput>
   > {
     const editor = await vscode.window.activeTextEditor;
 
-    let input;
+    let query;
 
     if (!editor) {
       const userInputOptions = {
         prompt: nls.localize('parameter_gatherer_enter_soql_query')
       } as vscode.InputBoxOptions;
-      input = await vscode.window.showInputBox(userInputOptions);
+      query = await vscode.window.showInputBox(userInputOptions);
     } else {
       const document = editor.document;
       if (editor.selection.isEmpty) {
         const userInputOptions = {
           prompt: nls.localize('parameter_gatherer_enter_soql_query')
         } as vscode.InputBoxOptions;
-        input = await vscode.window.showInputBox(userInputOptions);
+        query = await vscode.window.showInputBox(userInputOptions);
       } else {
-        input = document.getText(editor.selection);
+        query = document.getText(editor.selection);
       }
     }
 
-    input = input!.replace('[', '').replace(']', '');
-    return input ? { type: 'CONTINUE', data: { input } } : { type: 'CANCEL' };
+    query = query!.replace('[', '').replace(']', '');
+    return query ? { type: 'CONTINUE', data: { query } } : { type: 'CANCEL' };
+  }
+}
+
+export enum ApiType {
+  REST,
+  Tooling
+}
+
+export interface QueryInput {
+  query: string;
+}
+export interface ApiInput {
+  api: ApiType;
+}
+
+export class SelectApi implements ParametersGatherer<ApiInput> {
```

Comment:
Actually, this new gatherer seems very specific for this use case -- so there is little reuse. Is it possible to just fold it into the previous one instead of composing them with the `CompositeParameterGatherer`?

Reply:
@vazexqi No problem, I will make that change. For learning purposes, why is it better to fold it into the other gatherer?

Reply:
_This is a minor issue hence I didn't even notice it on first review_

It's mostly whether there is reuse for a gatherer. It seem (and I could be wrong) that the gatherer to select Rest vs Tooling API doesn't have any reuse outside of this case. Do you know of any other command where we might want to use it again (maybe the data commands?). If there is, then let's leave it as-is.

Another reason why we might keep it as-is is for testing. If you find it easier to test it if you separate it, it's fine to keep it this way too.

Reply:
Gotcha. The commands where I see the --usetoolingapi flag are the `sfdx force:data:record:*` commands, and the `sfdx force:schema:sobject:describe` command. Are there plans to implement those as vscode commands?

Reply:
`force:data:record:*` would not likely be surfaced as a top-level command using this UX mechanism – there would be too much typing involved with the current UX and is too error-prone. If we surface this we would have a UI layer on top.

`force:schema:sobject:describe` would also likely not be surfaced at the top-level as a command. 

I'm following the philosophy of the Git integration in VS Code. Git has many CLI commands. But not all need to be surfaced. If the user needs those commands, they can still bounce to the CLI as needed.

Reply:
@vazexqi sounds good! These changes are ready for another review